File: src/mouse.ts
import { native } from "./native";

export enum MouseEventType {
  Press = 0,
  Release = 1,
  Move = 2,
  Scroll = 3,
}

export enum MouseButton {
  Left = 1,
  Middle = 2,
  Right = 3,
  ScrollUp = 4,
  ScrollDown = 5,
}

export interface MouseEvent {
  eventType: MouseEventType;
  button: MouseButton;
  x: number;
  y: number;
  windowHandle: number;
}

type MousePressCallback = (event: MouseEvent) => void;
type MouseReleaseCallback = (event: MouseEvent) => void;
type MouseMoveCallback = (event: MouseEvent) => void;
type MouseScrollCallback = (event: MouseEvent) => void;

export class Mouse {
  private pressCallbacks: MousePressCallback[] = [];
  private releaseCallbacks: MouseReleaseCallback[] = [];
  private moveCallbacks: MouseMoveCallback[] = [];
  private scrollCallbacks: MouseScrollCallback[] = [];

  constructor() {}

  /**
   * Register a callback for mouse button press events
   */
  onMousePress(callback: MousePressCallback): void {
    this.pressCallbacks.push(callback);
  }

  /**
   * Register a callback for mouse button release events
   */
  onMouseRelease(callback: MouseReleaseCallback): void {
    this.releaseCallbacks.push(callback);
  }

  /**
   * Register a callback for mouse move events
   */
  onMouseMove(callback: MouseMoveCallback): void {
    this.moveCallbacks.push(callback);
  }

  /**
   * Register a callback for mouse scroll events
   */
  onScroll(callback: MouseScrollCallback): void {
    this.scrollCallbacks.push(callback);
  }

  /**
   * Process all pending mouse events (called by App)
   * @internal
   */
  processEvents(
    onPress?: MousePressCallback,
    onRelease?: MouseReleaseCallback,
    onMove?: MouseMoveCallback,
    onScroll?: MouseScrollCallback
  ): void {
    // Collect all events first, then process
    const events: MouseEvent[] = [];
    
    while (native.hasMouseEvents()) {
      const eventTypePtr = new BigUint64Array(1);
      const buttonPtr = new Uint32Array(1);
      const xPtr = new Int32Array(1);
      const yPtr = new Int32Array(1);
      const windowHandlePtr = new BigUint64Array(1);

      native.getNextMouseEvent(
        eventTypePtr,
        buttonPtr,
        xPtr,
        yPtr,
        windowHandlePtr
      );

      const event: MouseEvent = {
        eventType: Number(eventTypePtr[0]) as MouseEventType,
        button: buttonPtr[0]! as MouseButton,
        x: xPtr[0]!,
        y: yPtr[0]!,
        windowHandle: Number(windowHandlePtr[0]),
      };
      
      events.push(event);
    }
    
    // Skip redundant move events - only keep the last one
    const processedEvents: MouseEvent[] = [];
    let lastMoveEvent: MouseEvent | null = null;
    
    for (const event of events) {
      if (event.eventType === MouseEventType.Move) {
        lastMoveEvent = event; // Keep updating to latest
      } else {
        // Non-move event - flush any pending move event first
        if (lastMoveEvent) {
          processedEvents.push(lastMoveEvent);
          lastMoveEvent = null;
        }
        processedEvents.push(event);
      }
    }
    
    // Don't forget the last move event
    if (lastMoveEvent) {
      processedEvents.push(lastMoveEvent);
    }
    
    // Now process the optimized event list
    for (const event of processedEvents) {
      // Call per-window callback first if provided
      if (event.eventType === MouseEventType.Press && onPress) {
        onPress(event);
      } else if (event.eventType === MouseEventType.Release && onRelease) {
        onRelease(event);
      } else if (event.eventType === MouseEventType.Move && onMove) {
        onMove(event);
      } else if (event.eventType === MouseEventType.Scroll && onScroll) {
        onScroll(event);
      }

      // Call global callbacks
      if (event.eventType === MouseEventType.Press) {
        this.pressCallbacks.forEach((cb) => cb(event));
      } else if (event.eventType === MouseEventType.Release) {
        this.releaseCallbacks.forEach((cb) => cb(event));
      } else if (event.eventType === MouseEventType.Move) {
        this.moveCallbacks.forEach((cb) => cb(event));
      } else if (event.eventType === MouseEventType.Scroll) {
        this.scrollCallbacks.forEach((cb) => cb(event));
      }
    }
  }

  /**
   * Get the current mouse position
   */
  getPosition(): { x: number; y: number } {
    const xPtr = new Int32Array(1);
    const yPtr = new Int32Array(1);
    native.getMousePosition(xPtr, yPtr);
    return { x: xPtr[0]!, y: yPtr[0]! };
  }

  /**
   * Clear all pending mouse events
   */
  clear(): void {
    native.clearMouseEvents();
  }
}

File: src/native/mouse.zig
const std = @import("std");
const clib = @import("c.zig");
const c = clib.c;

const window = @import("window.zig");

// Mouse event types
pub const MouseEventType = enum(c_int) {
    Press = 0,
    Release = 1,
    Move = 2,
    Scroll = 3,
};

// Mouse buttons
pub const MouseButton = enum(c_int) {
    Left = 1,
    Middle = 2,
    Right = 3,
    ScrollUp = 4,
    ScrollDown = 5,
};

// Mouse event structure
pub const MouseEvent = extern struct {
    event_type: c_int, // MouseEventType
    button: c_int, // MouseButton (0 for move events)
    x: c_int,
    y: c_int,
    window_handle: c.Window,
};

// Mouse event queue (circular buffer)
const MAX_MOUSE_EVENTS = 128;
var mouse_event_queue: [MAX_MOUSE_EVENTS]MouseEvent = undefined;
var mouse_event_start: usize = 0;
var mouse_event_end: usize = 0;
var mouse_event_count: usize = 0;

// Track current mouse position per window
var last_mouse_x: c_int = 0;
var last_mouse_y: c_int = 0;

pub export fn handleMousePress(win: c.Window, button: c_uint, x: c_int, y: c_int) void {
    if (mouse_event_count >= MAX_MOUSE_EVENTS) return;

    var event: MouseEvent = undefined;
    event.event_type = @intFromEnum(MouseEventType.Press);
    event.button = @intCast(button);
    event.x = x;
    event.y = y;
    event.window_handle = win;

    mouse_event_queue[mouse_event_end] = event;
    mouse_event_end = (mouse_event_end + 1) % MAX_MOUSE_EVENTS;
    mouse_event_count += 1;

    last_mouse_x = x;
    last_mouse_y = y;
}

pub export fn handleMouseRelease(win: c.Window, button: c_uint, x: c_int, y: c_int) void {
    if (mouse_event_count >= MAX_MOUSE_EVENTS) return;

    var event: MouseEvent = undefined;
    event.event_type = @intFromEnum(MouseEventType.Release);
    event.button = @intCast(button);
    event.x = x;
    event.y = y;
    event.window_handle = win;

    mouse_event_queue[mouse_event_end] = event;
    mouse_event_end = (mouse_event_end + 1) % MAX_MOUSE_EVENTS;
    mouse_event_count += 1;

    last_mouse_x = x;
    last_mouse_y = y;
}

pub export fn handleMouseMove(win: c.Window, x: c_int, y: c_int) void {
    if (mouse_event_count >= MAX_MOUSE_EVENTS) return;

    var event: MouseEvent = undefined;
    event.event_type = @intFromEnum(MouseEventType.Move);
    event.button = 0; // No button for move events
    event.x = x;
    event.y = y;
    event.window_handle = win;

    mouse_event_queue[mouse_event_end] = event;
    mouse_event_end = (mouse_event_end + 1) % MAX_MOUSE_EVENTS;
    mouse_event_count += 1;

    last_mouse_x = x;
    last_mouse_y = y;
}

pub export fn handleMouseScroll(win: c.Window, button: c_uint, x: c_int, y: c_int) void {
    if (mouse_event_count >= MAX_MOUSE_EVENTS) return;

    var event: MouseEvent = undefined;
    event.event_type = @intFromEnum(MouseEventType.Scroll);
    event.button = @intCast(button); // 4 = up, 5 = down
    event.x = x;
    event.y = y;
    event.window_handle = win;

    mouse_event_queue[mouse_event_end] = event;
    mouse_event_end = (mouse_event_end + 1) % MAX_MOUSE_EVENTS;
    mouse_event_count += 1;
}

pub export fn hasMouseEvents() bool {
    return mouse_event_count > 0;
}

pub export fn getNextMouseEvent(
    event_type: *c_int,
    button: *c_int,
    x: *c_int,
    y: *c_int,
    window_handle: *c_ulong,
) bool {
    if (mouse_event_count == 0) return false;

    const event = mouse_event_queue[mouse_event_start];
    mouse_event_start = (mouse_event_start + 1) % MAX_MOUSE_EVENTS;
    mouse_event_count -= 1;

    event_type.* = event.event_type;
    button.* = event.button;
    x.* = event.x;
    y.* = event.y;
    window_handle.* = event.window_handle;

    return true;
}

pub export fn clearMouseEvents() void {
    mouse_event_start = 0;
    mouse_event_end = 0;
    mouse_event_count = 0;
}

pub export fn getMousePosition(x: *c_int, y: *c_int) void {
    x.* = last_mouse_x;
    y.* = last_mouse_y;
}
